core/src/main/scala/kafka/log/LogCleanerManager.scala

```diff
@@ -601,6 +601,9 @@ private[log] object LogCleanerManager extends Logging {
       // the active segment is always uncleanable
       Option(log.activeSegment.baseOffset),
 
+      // we do not want to clean past the high watermark
+      Option(log.highWatermark),
```

Comment:
I agree this probably makes sense. It is surprising that it was not there already. Was there a specific reason related to this patch that you decided to do it here? An alternative by the way is to replace the high watermark and first unstable offset here with `lastStableOffset`, which would represent the lower bound of the two.

Reply:
This would cause issues during tests where if we didn't update the highWatermark and tried to compact (and update the logStartOffset) we would throw an exception since we can't update the logStartOffset higher than the HW. This made me think we also shouldn't compact beyond the HW. I'm good with using `lastStableOffset`

Reply:
In practice, we do not allow the log start offset to get ahead of the high watermark, so I am not sure that is a real issue. Still, I cannot think of a strong reason to allow cleaning above the high watermark, so maybe it is simpler to prevent it.

Reply:
I switched this to `lastStableOffset`.